httpserver/dns_text_records.cpp

```diff
@@ -0,0 +1,194 @@
+            new_history.push_back(pow_difficulty_t{timestamp, difficulty});
+        }
+        return new_history;
+    } catch (const std::exception& e) {
+        LOKI_LOG(warn, "JSON parsing of PoW data failed: {}", e.what());
+        ec = std::make_error_code(std::errc::bad_message);
+        return new_history;
+    }
+}
+
+static std::string query_latest_version() {
+
+    std::error_code ec;
+    const std::string version_str = get_dns_record(LATEST_VERSION_URL, ec);
+
+    if (ec) {
+        return "";
+    }
+
+    return version_str;
+
+}
+
+struct version_t {
+    int major;
+    int minor;
+    int patch;
+};
+
+static bool is_old_version(version_t ours, version_t latest) {
+
+    if (ours.major > latest.major) {
+        return false;
+    }
+
+    if (ours.major < latest.major) {
+        return true;
+    }
+
+    // === the same major version ===
+
+    if (ours.minor > latest.minor) {
+        return false;
+    }
+
+    if (ours.minor < latest.minor) {
+        return true;
+    }
+
+    // === the same minor version ===
+
+    if (ours.patch >= latest.patch) {
+        return false;
+    } else {
+        return true;
+    }
+
+}
+
+static bool parse_version(const std::string& str, version_t& version_out) {
+    std::vector<std::string> strs;
+    strs.reserve(3);
+    boost::split(strs, str, boost::is_any_of("."));
+    if (strs.size() != 3) {
+        LOKI_LOG(warn, "Invalid format for the Storage Server version!");
+        return false;
+    }
+
+    try {
+        version_out.major = std::stoi(strs[0]);
+        version_out.minor = std::stoi(strs[1]);
+        version_out.patch = std::stoi(strs[2]);
+    } catch (const std::exception& e) {
+        LOKI_LOG(warn, "Invalid format for the Storage Server version! Error: {}", e.what());
+        return false;
+    }
+
+    return true;
+}
+
+
+void check_latest_version() {
+
+    const auto latest_version_str = query_latest_version();
+
+    if (latest_version_str.empty()) {
+        LOKI_LOG(warn, "Failed to retrieve or parse the latest version number from DNS record");
+        return;
+    }
+
+    version_t latest_version;
+    if (!parse_version(latest_version_str, latest_version)) {
+        LOKI_LOG(warn, "Could not parse the latest version: {}", latest_version_str);
+        return;
+    }
+
+    // Note: we shouldn't have to parse our version every time, but we don't care about performance here
+    version_t our_version;
+    if (!parse_version(STORAGE_SERVER_VERSION_STRING, our_version)) {
+        LOKI_LOG(warn, "Could not parse our version: {}", STORAGE_SERVER_VERSION_STRING);
+        return;
+    }
+
+    if (is_old_version(our_version, latest_version)) {
+        LOKI_LOG(warn,
+                 "You are using an outdated version of the storage server "
+                 "({}), please update to {}!",
+                 STORAGE_SERVER_VERSION_STRING, latest_version_str);
+    } else {
+        LOKI_LOG(
+            debug,
```

Comment:
Can we assume that the lack of warning means we are on the latest version?
It might get too spammy otherwise to print a message every 10minutes to say that everything is fine...
![download](https://user-images.githubusercontent.com/40749766/62748064-8cb65400-ba9a-11e9-951c-09de426a27db.jpeg)

Reply:
It is on debug level, which is won't be shown to the user, so should be OK?

Reply:
Yeah cool :)